File: examples/segment_tree.range_sum_query.test.cpp
#define PROBLEM "https://onlinejudge.u-aizu.ac.jp/courses/library/3/DSL/all/DSL_2_B"
#include <iostream>
#include "examples/segment_tree.hpp"
#include "examples/monoids.hpp"
#include "examples/macros.hpp"
using namespace std;

int main() {
    int n, q; cin >> n >> q;
    segment_tree<plus_monoid> segtree(n);
    REP (i, n) {
        segtree.point_set(i, 0);
    }
    REP (i, q) {
        int com, x, y; cin >> com >> x >> y;
        -- x;
        if (com == 0) {
            segtree.point_set(x, segtree.range_concat(x, x + 1) + y);
        } else if (com == 1) {
            cout << segtree.range_concat(x, y) << endl;
        }
    }
    return 0;
}

File: examples/csharpsx/Tests/segment_tree.range_sum_query.test.cs
using System;
using System.Linq;
using Library;

// verification-helper: PROBLEM https://onlinejudge.u-aizu.ac.jp/courses/library/3/DSL/all/DSL_2_B
class range_sum_query
{
    static void Main()
    {
        var nq = Console.ReadLine().Split().Select(int.Parse).ToArray();
        var (n, q) = (nq[0], nq[1]);
        SegmentTree<int> segTree = new SegmentTree<int>(n, 0, (x, y) => x + y);

        for (int i = 0; i < q; i++)
        {
            var query = Console.ReadLine().Split().Select(int.Parse).ToArray();
            if (query[0] == 0)
            {
                segTree[query[1] - 1] += query[2];
            }
            else
            {
                Console.WriteLine(segTree[(query[1] - 1)..query[2]]);
            }
        }
    }
}
